racket: refactor and add features, 8.14 -> 8.15, mark Darwin as `badPlatforms`

rc-zb opened the pull request:
1. Refactor `racket` and `racket-minimal`.

   - Reduce the closure size of `racket-minimal` by removing redundant dependencies. (fixing #246902)
   - Unify `pname` to `racket`.
   - Clean code.

2. Add features to Racket derivations.

   - Add `writeScript[Bin]`.
   - Add some package tests.
   - Add `updateScript`.

3. Update racket from 8.14 to 8.15 (alternative to #360765, a trivial update on the version and hashes only).

4. Sadly, both this and #360765 (and even the original recipe -- see #372197) seem to fail to build on Darwin. I have to mark it as `badPlatforms` since I don't have a macOS device to test it.

5. Add myself to the their maintainer list, which was empty until now.
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Limitations

- I guess it could be no more possible to extend the minimal distribution to a full one with `raco pkg` after installation, for lack of runtime dependencies. I don't know if this is part of the public interface of which all breaking changes must be documented.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
+    export DYLD_FALLBACK_LIBRARY_PATH="${finalAttrs.LD_LIBRARY_PATH}"
+  '';
+
+  enableParallelBuilding = false;
+
```
```suggestion
```
thats the default

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
+    unset AR
+    for f in src/lt/configure src/cs/c/configure src/bc/src/string.c; do
+      substituteInPlace "$f" \
+        --replace /usr/bin/uname ${coreutils}/bin/uname \
+        --replace /bin/cp ${coreutils}/bin/cp \
+        --replace /bin/ln ${coreutils}/bin/ln \
+        --replace /bin/rm ${coreutils}/bin/rm \
+        --replace /bin/true ${coreutils}/bin/true
```
This should be using --replace-fail

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
+    # The configure script forces using `libtool -o` as AR on Darwin. But, the
+    # `-o` option is only available from Apple libtool. GNU ar works here.
+    substituteInPlace src/ChezScheme/zlib/configure \
+        --replace 'ARFLAGS="-o"' 'AR=ar; ARFLAGS="rc"'
```
same here

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
+    }
+  ) { hash = "sha256-/BhnyIw4QQ2aMKW8FDJlsalSW3pTLqP74CXJ15td5so="; };
+
+  libraries = [ ncurses ];
+
```
```suggestion
```
just write that out, it is not worth it to have the complexity.

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
+  src = lib.makeOverridable (
+    { hash }:
+    fetchurl {
+      url = "https://mirror.racket-lang.org/installers/${finalAttrs.version}/${finalAttrs.pname}-${finalAttrs.version}-src.tgz";
```
```suggestion
      url = "https://mirror.racket-lang.org/installers/${finalAttrs.version}/racket-minimal-${finalAttrs.version}-src.tgz";
```

SuperSandro2000 commented on pkgs/development/interpreters/racket/minimal.nix:
```diff
@@ -1,25 +1,137 @@
-  };
+  version = "8.14";
+
+  src = lib.makeOverridable (
+    { hash }:
+    fetchurl {
```
Now we can also remove this

SuperSandro2000 commented on pkgs/development/interpreters/racket/default.nix:
```diff
@@ -1,209 +1,109 @@
+
+    FONTCONFIG_FILE = makeFontsConf { fontDirectories = [ freefont_ttf ]; };
+
+    libraries =
```
Why go the indirection of libraries and not straight use LD_LIBRARIES_PATH.

Also do we really need to use LD_LIBRARIES_PATH? It is really hacky, patchelf would be nicer.

rc-zb replied:
Actually I don't know what this means; they are just left by the previous maintainers. I'll give a try for alternatives.

rc-zb replied:
The builder of Racket packages can't find `libfontconfig.so.1` in absence of `LD_LIBRARY_PATH`, even if `fontconfig` is put in `buildInputs`. `autoPatchelfHook` works only at the fixup phase and does nothing helpful.

SuperSandro2000 commented on pkgs/development/interpreters/racket/default.nix:
```diff
@@ -1,209 +1,109 @@
+    pname = "racket";
+    version = "8.14";
+
+    src = previousAttrs.src.override {
+      hash = "sha256-qrjMDbM27S04KAPHCK1VqV/FKkQ2yRL2FvfEnUhFriw=";
+    };
```
Relying on finalAttrs.pname here is really none transparent and hard to understand.

rc-zb committed:
racket and racket-minimal: refactor

- Remove redundant dependencies and patch. (fix #246902)
- Unify `pname` to `racket`.
- Clean code.

rc-zb committed:
racket: add writeScript[Bin]

rc-zb committed:
racket: add tests

rc-zb committed:
racket: add updateScript

rc-zb committed:
racket: 8.14 -> 8.15

Release note: https://blog.racket-lang.org/2024/11/racket-v8-15.html

Close #360765.

rc-zb committed:
racket: mark Darwin as badPlatforms

rc-zb commented:
> I haven't reviewed this PR in detail, but I wrote in [#209660 (comment)](https://github.com/NixOS/nixpkgs/issues/209660#issuecomment-1621942022) about how we handle this in Guix.

Thanks for directing me to the discussion. I do replace the `LD_LIBRARY_PATH` wrapper with `config.rktd` entries so that it finds the "necessary" foreign libraries cleanly. However, due to how Nix works and what it favors, it is neither very feasible nor expected that Racket just takes the library paths from the global environment. On the other hand, to facilitate things in a Nix shell, we can make use of the [setup hook](https://nixos.org/manual/nixpkgs/unstable/#ssec-setup-hooks) mechanism as Guile [does](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/guile/setup-hook-3.0.sh). I decide to leave them to a further PR, where Racket packages would get built with Nix.

pillowtrucker commented:
@rc-zb @SuperSandro2000 did this PR change the distribution of `.boot` files from standalone to embedded in executable? I can no longer build the rust `racket-sys` bindings due to those missing and I can't actually find them manually any more
